### ISSUE TITLE
While running terraform destory -target=module.somemodule a variable is not being recognized

### ISSUE THREAD

**Issue by yairzaslavsky** [hashicorp/terraform#9042]
### Terraform Version
Terraform version 0.74

### Affected Resource(s)
I assume this has to be with terraform's core (module /variable mangement)

### Terraform Configuration Files

directory structure:
-- main.tf
--variables.tf
 |---> somemodule
          |
          | --- variables.tf
          | ---- resources.tf
 | --> somemodule2
          |
          |  -- variables.tf
          | -- resources.tf

variables.tf in the root dir includes:

variable "my_instances" {
  type = "list"
  description = "List of babylon instances"
  default = ["instance1", "instance2"]
}

main.tf includes: 

module "somemodule" {
  source = "./somemodule"
  my_instances = "${var.my_instances}"
}

somemodule/reosurces.tf includes:

resource "aws_s3_bucket" "some_buckets" {
  count = "${length(var.my_instances)}"
  bucket = "prefix-${element(var.my_instances, count.index)}"
  acl = "private"
}

somemodule2 includes another aws resource that is based on the same "my_instances" variable.

terraform apply -target=module.somemodule   

builds the infrastructure correctly.

terraform destroy -target=module.somemodule 

fails to destroy the infrastructure correctly, and i get the error:

Error configuring: 2 error(s) occurred:

* 1:10: unknown variable accessed: var.my_instances in:       

### Debug Output

https://gist.github.com/yairzaslavsky/b984a89b1f08e395c157d085f6476cab

### Expected Behavior
What should have happened?

### Actual Behavior
What actually happened?

### Steps to Reproduce

1. terraform apply -target=module.somemodule
2. terraform destroy -target=module.somemodule

### Important Factoids
The ACS for s3 is private

**Comment by yairzaslavsky** [comment]
Unfortunately my ASCII tree did not turn out that well,
I meant that both module directories have a resources.tf file and a variables.tf file

**Comment by bdashrad** [comment]
I believe i'm encountering a similar error with `terraform plan -destroy -target=module.somemodule -var-file=./vars.tfvars`

```
$ terraform plan --target=module.brad-forms --var-file=dev.tfvars -destroy
Error configuring: 1 error(s) occurred:

* 1:3: unknown variable accessed: var.region in:

${var.region}
```

```
$ terraform -v
Terraform v0.7.4
```

**Comment by bdashrad** [comment]
it appears to not affect variables with defaults set in the module itself. It also doesn't seem to affect all variables, which is very strange....

**Comment by Jonnymcc** [comment]
I am seeing similar behavior in 0.7.4 as well.

**Comment by pccowboy** [comment]
this does not appear to be tied specifically to "destroy" in 0.7.9  - I can execute 
  terraform plan -refresh=true

but not
  terraform plan -refresh=true -target=aws_db_option_group.speyside-mysql-5-7

the error is 
```
Error configuring: 1 error(s) occurred:

* 1:25: unknown variable accessed: var.env in:

${lookup(var.aws_region, var.env)}
```

**Comment by marionettist** [comment]
Seeing this issue in 0.7.13 and 0.8.2

**Comment by MikeMcMahon** [comment]
running into this issue with 0.8.2 adding -Xlegacy-graph works for the time being

**Comment by bdashrad** [comment]
This happens seemingly without rhyme or reason while doing targeted runs in terraform. It's very frustrating.

**Comment by MikeMcMahon** [comment]
Pretty consistent with us. We have several nested modules that simply don't work after moving to 0.8.2. Redefining the variables works within the modules but that defeats the purpose of defining them once at an outer scope and isn't reasonable for us to modify every single module. Plan works without the target though. :/

**Comment by ste-bah** [comment]
Also happening in 0.8.4 consistently for us when using target and seems to jump to random variables,  full plan works fine.  This was fine in 0.7.10 so it might be something to do with the new graphs.  Using the **-Xlegacy-graph** works when using target.

**Comment by eyalzek** [comment]
Geting the same on 0.8.5 when planning on a nested module:

```
Error running plan: 2 error(s) occurred:

* 1:3: unknown variable accessed: var.region in:

${var.region}
* 1:3: unknown variable accessed: var.region in:

${var.region}
```

As mentioned previously, running with `-Xlegacy-graph` solves that problem. I didn't see this issue on 0.7.8.
I remember facing the same problems when upgrading our codebase to 0.7X and opening a ticket, seems like the parent ticket for that is still open:
https://github.com/hashicorp/terraform/issues/8146

**Comment by schinivision** [comment]
I also ran into that issue... Terraform code worked until 0.7.13 perfectly.
Wanted to upgrade to 0.8.5 and ran into the issue

* 1:3: unknown variable accessed: var.aws_access_key in:

${var.aws_access_key}
* 1:3: unknown variable accessed: var.aws_access_key in:

${var.aws_access_key}

This error occurs on all Versions beginning with 0.8.0 up to current... stepping back to 0.7.13 everything works like a charm.

I also tried the -Xlegacy-graph option... helped that it is possible to plan again... But I don't think this is an intended behaviour

**Comment by MikeMcMahon** [comment]
has there been any movement or traction on this issue?

**Comment by ste-bah** [comment]
I don't believe so I'm still getting issues in 0.8.8 but using Xlegacy-graph option still works for now - we are also using nested modules.

**Comment by eyalzek** [comment]
Now that v0.9 is released, is `-Xlegacy-graph` actually deprecated?

**Comment by stefanasseg** [comment]
Any progress on this? This currently hinders us to upgrade our Terraform version. Thank you!

**Comment by MikeMcMahon** [comment]
this is still a major issue for us and prevents us from upgrading to the latest terraform, being version locked sucks as we're missing out on new features :(

**Comment by ste-bah** [comment]
Just an update - I've updated to 0.10.2 and this seems to work now  on our test environment.
